File: heap.c
#include "heap.h"
#include <stdlib.h>
#include <stddef.h>

typedef struct heap_value_s heap_value_t;

typedef struct heap_node_s
{
    heap_value_t* heap_value;
    struct heap_node_s* left;
    struct heap_node_s* right;
    struct heap_node_s* top;
}heap_node_t;

struct heap_value_s
{
    heap_node_t* node;
    void* value;
};

typedef struct
{
    struct heap_node_s* entry;
    int length;
    bool(*compare)(void* A,void* B);
}heap_t;

/* function predefines */
heap_node_t* heap_get_top(heap_t* heap,int n,bool* is_left);
heap_node_t* heap_adjust_backward(heap_t* heap,heap_node_t* node);
heap_node_t* heap_adjust_forward(heap_t* heap,heap_node_t* node);
void heap_free_node(heap_node_t* node,void(*free_value)(void* value,void* ctx), void* ctx,bool propagate);
heap_node_t* heap_new_node(heap_value_t* heap_value);
heap_value_t* heap_new_value(void* value);

/* public functions */

heap_handle_t heap_create(bool(*compare)(void* A,void* B))
{
    if(compare == NULL)
        return NULL;
    heap_t* heap = malloc(sizeof(heap_t));
    if(!heap)
        return NULL;
    heap->compare = compare;
    heap->entry = NULL;
    heap->length = 0;
    return (heap_handle_t)heap;
}

void heap_free(heap_handle_t handle,void(*free_value)(void* value,void* ctx),void* ctx)
{
    if(handle == NULL)
        return;
    heap_t* heap = (heap_t*)handle;
    heap_free_node(heap->entry,free_value,ctx,true);
    free(heap);
}

heap_value_handle_t heap_add(heap_handle_t handle,void* value)
{
    if(!handle)
        return NULL;
    heap_t* heap = (heap_t*)handle;
    heap_value_t* heap_value = heap_new_value(value);
    if(!heap_value)
        return NULL;
    heap_node_t* new_node = heap_new_node(heap_value);
    if(!new_node)
    {
        free(heap_value);
        return NULL;
    }
    if(heap->entry == NULL)
    {
        heap->entry = new_node;
        heap->length ++;
        return (void*)heap_value;
    }
    bool is_left;
    heap_node_t* node = heap_get_top(heap,++heap->length,&is_left);
    if(is_left)
        node->left = new_node;
    else
        node->right = new_node;
    new_node->top = node;
    heap_adjust_backward(heap,new_node);
    return (void*)heap_value;
}

bool heap_delete(heap_handle_t handle,heap_value_handle_t heap_value_handle)
{
    if(!handle)
        return false;
    heap_t* heap = (heap_t*)handle;
    if(!heap_value_handle)
        return false;
    heap_value_t* heap_value = (heap_value_t*)heap_value_handle;
    heap_node_t* node = heap_value->node;
    if(!node)
        return false;
    if(heap->length == 1)
    {
        /* free the only node */
        heap_free_node(node,NULL,NULL,false);
        heap->entry = NULL;
        heap->length --;
        return true;
    }
    bool is_left;
    heap_node_t* tail_top_node = heap_get_top(heap,heap->length,&is_left);
    heap_node_t* tail_node = is_left?tail_top_node->left:tail_top_node->right;
    /* switch node & tail_node value */
    heap_value_t* temp_heap_value = node->heap_value;
    node->heap_value = tail_node->heap_value;
    node->heap_value->node = node;
    tail_node->heap_value = temp_heap_value;
    /* redundant */
    // tail_node->heap_value->node = tail_node;
    /* free the tail node */
    heap_free_node(tail_node,NULL,NULL,false);
    heap->length --;
    if(is_left)
        tail_top_node->left = NULL;
    else
        tail_top_node->right = NULL;
    if(node != tail_node)
    {
        node = heap_adjust_backward(heap,node);
        heap_adjust_forward(heap,node);
    } 
    return true;
}

bool heap_update(heap_handle_t handle,heap_value_handle_t value_handle)
{
    if(!handle)
        return false;
    heap_t* heap = (heap_t*)handle;
    if(!value_handle)
        return false;
    heap_value_t* heap_value = (heap_value_t*)value_handle;
    heap_node_t* node = heap_value->node;
    if(!node)
        return false;
    node = heap_adjust_backward(heap,node);
    heap_adjust_forward(heap,node);
    return true;
}

void* heap_pop(heap_handle_t handle)
{
    if(!handle)
        return NULL;
    heap_t* heap = (heap_t*)handle;
    void* result = NULL;
    if(heap->entry != NULL)
        result = heap->entry->heap_value->value;
    if(heap->length == 1)
    {
        heap->length--;
        /* free the only node */
        heap_free_node(heap->entry,NULL,NULL,false);
        heap->entry = NULL;
    }
    else if(heap->length > 1)
    {
        bool is_left;
        heap_node_t* tail_top_node = heap_get_top(heap,heap->length,&is_left);
        heap_node_t* tail_node = is_left?tail_top_node->left:tail_top_node->right;
        heap_value_t* temp_heap_value = heap->entry->heap_value;
        /* switch entry & tail value */
        heap->entry->heap_value = tail_node->heap_value;
        heap->entry->heap_value->node = heap->entry;
        tail_node->heap_value = temp_heap_value;
        /* redundant */
        // tail_node->heap_value->node = tail_node;
        /* free the tail node */
        heap_free_node(tail_node,NULL,NULL,false);
        if(is_left)
            tail_top_node->left = NULL;
        else
            tail_top_node->right = NULL;
        heap->length--;
        heap_adjust_forward(heap,heap->entry);
    }
    return result;
}

void* heap_get(heap_handle_t handle)
{
    if(!handle)
        return NULL;
    heap_t* heap = (heap_t*)handle;
    void* result = NULL;
    if(heap->entry != NULL)
        result = heap->entry->heap_value->value;
    return result;
}

int heap_get_length(heap_handle_t handle)
{
    if(!handle)
        return -1;
    heap_t* heap = (heap_t*)handle;
    return heap->length;
}

/* private functions */
heap_node_t* heap_get_top(heap_t* heap,int n,bool* is_left)
{
    if(n==1)
        return NULL;
    heap_node_t* node = heap->entry;
    int layers = 32 - __builtin_clz(n);
    for(int i=layers-2;i>0;i--)
    {
        if((n>>i)&1)
        {
            node = node->right;
        }
        else
        {
            node = node->left;
        }
    }
    *is_left = !(n&1);
    return node;
}

heap_node_t* heap_adjust_backward(heap_t* heap,heap_node_t* node)
{
    if(node->top == NULL)
        return node;
    if(heap->compare(node->top->heap_value->value,node->heap_value->value))
    {
        heap_value_t* temp_heap_value = node->heap_value;
        node->heap_value = node->top->heap_value;
        node->heap_value->node = node;
        node->top->heap_value = temp_heap_value;
        node->top->heap_value->node = node->top;
        return heap_adjust_backward(heap,node->top);
    }
    else
    {
        return node;
    }
}

heap_node_t* heap_adjust_forward(heap_t* heap,heap_node_t* node)
{
    heap_node_t* target_node = NULL;
    if(node->left != NULL && node->right != NULL)
    {
        heap_node_t* less_node = NULL;
        if(heap->compare(node->left->heap_value->value,node->right->heap_value->value))
        {
            less_node = node->right;
        }
        else
        {
            less_node = node->left;
        }
        if(heap->compare(node->heap_value->value,less_node->heap_value->value))
        {
            target_node = less_node;
        }
    }
    else if(node->left != NULL && node->right == NULL)
    {
        if(heap->compare(node->heap_value->value,node->left->heap_value->value))
        {
            target_node = node->left;
        }
    }
    else if(node->left == NULL && node->right != NULL)
    {
        if(heap->compare(node->heap_value->value,node->right->heap_value->value))
        {
            target_node = node->right;
        }
    }
    if(target_node != NULL)
    {
        heap_value_t* temp_heap_value = node->heap_value;
        node->heap_value = target_node->heap_value;
        node->heap_value->node = node;
        target_node->heap_value = temp_heap_value;
        target_node->heap_value->node = target_node;
        return heap_adjust_forward(heap,target_node);
    }
    else
    {
        return node;
    }
}

void heap_free_node(heap_node_t* node,void(*free_value)(void* value,void* ctx), void* ctx,bool propagate)
{
    if(node == NULL)
        return;
    
    heap_node_t* left = node->left;
    heap_node_t* right = node->right;

    if(node->heap_value != NULL)
    {
        if(free_value != NULL)
            free_value(node->heap_value->value,ctx);
        free(node->heap_value);
    }
    free(node);
    
    if(propagate)
    {
        heap_free_node(left,free_value,ctx,propagate);
        heap_free_node(right,free_value,ctx,propagate);
    }
}

heap_node_t* heap_new_node(heap_value_t* heap_value)
{
    heap_node_t* node = malloc(sizeof(heap_node_t));
    if(!node)
        return NULL;
    node->left = NULL;
    node->right = NULL;
    node->top = NULL;
    node->heap_value = heap_value;
    heap_value->node = node;
    return node;
}

heap_value_t* heap_new_value(void* value)
{
    heap_value_t* heap_value = malloc(sizeof(heap_value_t));
    if(!heap_value)
        return NULL;
    heap_value->value = value;
    return heap_value;
}

File: Heap.mjs
// @ts-check

class HeapNode{
    /**
     * @type {HeapNode}
     */
    left = undefined;
    /**
     * @type {HeapNode}
     */
    right = undefined;
    /**
     * @type {HeapNode}
     */
    top = undefined;
    /**
     * 
     * @param {HeapValue} value 
     */
    constructor(value) {
        this.value = value;   
    }

    get dump(){
        return {
            value:this.value,
            left:this.left?.dump,
            right:this.right?.dump
        };
    }
}

export class HeapValue{
    /** @type {HeapNode} */
    node = undefined;
    value = undefined;
    /**
     * @param {any} value 
     */
    constructor(value) {
        this.value = value;
    }
}

export class Heap{
    /**
     * @type {HeapNode}
     */
    #entry = undefined;
    #length = 0;
    /**
     * @param {(A:any,B:any)=>boolean} compare the relation of bottom->top
     */
    constructor(compare) {
        /**
         * 
         * @param {HeapNode} n1 
         * @param {HeapNode} n2 
         * @returns {Boolean}
         */
        this.#compare = (n1,n2)=>{
            return compare(n1.value.value,n2.value.value);
        };
    }

    /**
     * 
     * @param {HeapValue} value 
     */
    add(value){
        let newNode = new HeapNode(value);
        value.node = newNode;
        if(this.#entry === undefined){
            this.#entry = newNode;
            this.#length ++;
            return;
        }
        let {node,branch} = this.#getTop(++this.#length);
        node[branch] = newNode;
        newNode.top = node;
        this.#adjustBackward(newNode);
    }

    /**
     * @param {HeapValue} value 
     */
    delete(value){
        let node = value.node;
        if(node === undefined){
            return false;
        }
        if(this.#length === 1){
            this.#entry = undefined;
            this.#length --;
            return true;
        }
        let tailTop = this.#getTop(this.#length);
        let tailNode = tailTop.branch==='left'?tailTop.node.left:tailTop.node.right;
        node.value = tailNode.value;
        node.value.node = node;
        this.#length --;
        if(tailTop.branch==='left'){
            tailTop.node.left = undefined;
        }else{
            tailTop.node.right = undefined;
        }
        if(node !== tailNode){
            node = this.#adjustBackward(node);
            this.#adjustForeward(node);
        }
        return true;
    }

    /**
     * @param {HeapValue} value 
     */
    adjust(value){
        let node = value.node;
        if(node === undefined){
            return false;
        }
        node = this.#adjustBackward(node);
        this.#adjustForeward(node);
        return true;
    }

    /**
     * @returns {HeapValue}
     */
    get(){
        return this.#entry?.value;
    }

    /**
     * @returns {HeapValue}
     */
    pop(){
        const result = this.#entry?.value;
        // TODO: remove entry
        if(this.#length === 1){
            this.#length --;
            this.#entry = undefined;
        }else if(this.#length > 1){
            let {node,branch} = this.#getTop(this.#length);
            this.#entry.value = node[branch].value;
            this.#entry.value.node = this.#entry;
            node[branch] = undefined;
            this.#length --;
            this.#adjustForeward(this.#entry);
        }
        if(result !== undefined){
            result.node = undefined;
        }
        return result;
    }

    /**
     * @param {HeapNode} A 
     * @param {HeapNode} B
     * @returns {boolean} 
     */
    #compare = (A,B)=>undefined;

    /**
     * 
     * @param {HeapNode} node the top node  
     * @returns {HeapNode}
     */
    #adjustForeward(node){
        let targetNode = undefined;
        if(node.left !== undefined && node.right === undefined){
            if(this.#compare(node,node.left)){
                targetNode = node.left;
            }
        }else if(node.left === undefined && node.right !== undefined){
            if(this.#compare(node,node.right)){
                targetNode = node.right;
            }
        }else if(node.left !== undefined && node.right !== undefined){
            let lessNode = undefined;
            if(this.#compare(node.left,node.right)){
                lessNode = node.right;
            }else{
                lessNode = node.left;
            }
            if(this.#compare(node,lessNode)){
                targetNode = lessNode;
            }
        }
        if(targetNode !== undefined){
            // exchange value instead of adjusting connection
            let tempValue = node.value;
            node.value = targetNode.value;
            node.value.node = node;
            targetNode.value = tempValue;
            targetNode.value.node = targetNode;
            return this.#adjustForeward(targetNode);
        }else{
            return node;
        }
    }

    /**
     * 
     * @param {HeapNode} node the bottom node 
     * @returns {HeapNode}
     */
    #adjustBackward(node){
        if(node.top === undefined){
            return node;
        }
        if(this.#compare(node.top,node)){
            // exchange value instead of adjusting connection
            let tempValue = node.value;
            node.value = node.top.value;
            node.value.node = node;
            node.top.value = tempValue;
            node.top.value.node = node.top;
            return this.#adjustBackward(node.top);
        }else{
            return node;
        }
    }

    /**
     * 
     * @param {Number} n starts at 1
     * @returns {{
     *      node:HeapNode,
     *      branch:String
     * }} 
     */
    #getTop(n){
        if(n===1){
            return undefined;
        }
        let node = this.#entry;
        const layers = 32 - Math.clz32(n);
        for(let i=layers-2;i>0;i--){
            if(((n>>>i)&1)===0){
                node = node?.left;
            }else{
                node = node?.right;
            }
        }
        return {node,branch:n&1?'right':'left'};
    }

    get dump(){
        return this.#entry?.dump;
    }

    get length(){
        return this.#length;
    }
}




